### PR TITLE
Add API endpoint: /v1/actions/execute-command

### LINKED ISSUE #8034: API endpoint: /v1/actions/execute-command
## Is your feature request related to a problem? Please describe.

As a user I'd like to let a particular endpoint execute a particular check/notification/event-command in context of a particular checkable.

Also I'd like to see:

* the exit code / kill signal / execution error
* output

## Describe the solution you'd like

* POST /v1/actions/execute-command
  * takes:
    * the same target checkables specification parameters as POST /v1/actions/schedule-downtime
    * a macro string `endpoint`
      * defaults to `$command_endpoint$`
      * must resolve to a valid endpoint name (!= "") on all target checkables
    * a string `command_type`
      * CheckCommand/EventCommand/NotificationCommand
      * defaults to EventCommand
    * a macro string `command`
      * depending on `command_type` defaults to either `$check_command$` or `$event_command$` or ""
      * must resolve to a valid command name (!= "") of `command_type` on all target checkables
    * a dictionary `macros`
      * defaults to {}
      * overrides resolved macros (including the ones in the other parameters), e.g. `{"http_ssl": false}`
    * (a macro string resolving to) a number `ttl`
    * a boolean (like `pretty`) `wait`
  * does (action-specific):
    1. Collects the target checkables like POST /v1/actions/schedule-downtime, but with the new permission `objects/execute/<type>`
    2. Resolves (and validates) all of the above macro strings
    3. Resolves the macros as if for letting a command endpoint execute a check command
    4. Generates one UUID per target checkable
    5. Computes an absolute deadline from the TTL
    6. Routes messages to the endpoints to run the commands on:
      * method: event::ExecuteCommand
      * params:
        * like the event::ExecuteCommand ones
        * a UUID string `source` – the UUID generated above
        * a timestamp number `deadline` – the deadline computed above
    7. If `wait` is true:
      * Waits for the command executions to finish
      * Returns a dictionary:
        * key: checkable name
        * value: dictionary – keys:
          * `error`: execution error if any
          * `signal`: termination signal if any
          * `exit`: exit code if any
          * `output`: output if could be started
          * `start`: start time (\*nix timestamp) if could be started
          * `end`: end time (\*nix timestamp) if could be started
    8. Otherwise: returns a dictionary with checkable name mapped to its UUID (202, not 200 OK)
* GET /v1/actions/execute-command
  * takes:
    * one or more UUID strings `execution` (AFAIK there are already some parameters in existing actions which may be specified multiple times, see there)
  * does (action-specific):
    1. Returns a sub-response per `execution` parameter (AFAIK there are already some existing actions which do that, see there):
      * If the execution doesn’t exist or is expired: 404
      * Else, if the execution finished: a dictionary as from POST `/v1/actions/execute-command?wait=1` below the checkable name
      * Else: a dictionary `{"pending": true}`
* event::ExecuteCommand `source` handling
  * Executes the command as usual, but instead of converting the output to a check result, send it as a event::ExecutedCommand cluster message:
    * `execution` – the UUID
    * attributes as from POST `/v1/actions/execute-command?wait=1` below the checkable name
  * Sends on rejection (accepts commands? hierarchy ok? etc.) also a such message with the reason as execution error
* event::ExecutedCommand handling
  * If not rejected (accepts commands? hierarchy ok? etc.), mark execution as finished

## Describe alternatives you've considered

#7709

## Additional context

@lippserd

### PR DESCRIPTION
**Pull request by Al2Klimov**
fixes #8034
closes #7709

### CODE REVIEW

**Comment by Al2Klimov** [comment]
@lippserd Whoever is going to actually implement this shall work based on this branch **and create a PR targeting this branch**. Also if the stubs aren't enough, just say so.

**Comment by Al2Klimov** [comment]
# Actual implementation guide

1. Instead of [this stub](https://github.com/Icinga/icinga2/pull/8040/commits/752df8ea283de41101406b10f74bfd78f6384440#diff-b3d41e16e7f7f292c6c5f10c8d26b72fR558):
    * Collect and validate the parameters `endpoint`, `command_type`, `command`, `macros`, `ttl` and `wait` described in #8034 (see other actions in the same file on how to do it)
    * Resolve the macro string ones (see [here](https://github.com/Icinga/icinga2/blob/fd4abff369c7268d1c61219713b73c3c02c4b22f/lib/methods/dummychecktask.cpp#L31-L42) on how to do it)
    * Ensure the objects with the resolved names exist (see [here](https://github.com/Icinga/icinga2/blob/fd4abff369c7268d1c61219713b73c3c02c4b22f/lib/remote/zone.cpp#L31-L33) on how to do it)
    * Validate the resolved macro strings not referring to config objects
    * Resolve the macros as if for letting a command endpoint execute a check command (see [here](https://github.com/Icinga/icinga2/blob/fd4abff369c7268d1c61219713b73c3c02c4b22f/lib/methods/pluginchecktask.cpp#L31-L36) and [here](https://github.com/Icinga/icinga2/blob/fd4abff369c7268d1c61219713b73c3c02c4b22f/lib/icinga/pluginutility.cpp#L21-L72) on how to do it)
    * Generate a UUID ([this one](https://github.com/Icinga/icinga2/blob/fd4abff369c7268d1c61219713b73c3c02c4b22f/lib/base/utility.cpp#L375-L383) is your friend)
    * Compute an absolute deadline from the TTL
    * Update `Checkable#executions` (see above) adding the new execution as a pending one and broadcast that update via event::UpdateExecutions (see [here](https://github.com/Icinga/icinga2/pull/8040/commits/9cb3779ec2dfc1623cf7b3ed727a5b8197396921#diff-333863a57a158a50c87772836c137c99R950-R953) for a stub and [here](https://github.com/Icinga/icinga2/blob/fd4abff369c7268d1c61219713b73c3c02c4b22f/lib/icinga/clusterevents.cpp#L172-L192) on how to do it)
    * Let the target endpoint run the command via event::ExecuteCommand (see [here](https://github.com/Icinga/icinga2/blob/fd4abff369c7268d1c61219713b73c3c02c4b22f/lib/icinga/checkable-check.cpp#L555-L584) on how to do it) and include the params `source` and `deadline` (see #8034)
    * See #8034 on how to handle the `wait` param
2. Instead of [this stub](https://github.com/Icinga/icinga2/pull/8040/commits/9cb3779ec2dfc1623cf7b3ed727a5b8197396921#diff-333863a57a158a50c87772836c137c99R952):
    * Check whether the sender is allowed to send this (see other handlers in the same file on how to do it) and update `Checkable#executions`
3. Instead of [this stub](https://github.com/Icinga/icinga2/pull/8040/commits/3f07dd6caeb15662ee7e33479d2480f8eed6062a#diff-2c5f14666a76d5fb4c06e876ec241673R96):
    * Check whether the sender is allowed to send this and execute the command (see the containing methods on how to do it)
    * ~Before actually executing, store what to do in `Checkable#pending_executions` just for the case of a node restart during execution~ @lippserd Forget the persistence of not yet executed commands on the target endpoints! If your command itself causes a restart...
    * Send an event::ExecutedCommand (see #8034) to the execution requestor
4. Instead of [this stub](https://github.com/Icinga/icinga2/pull/8040/commits/fc220bac69370019250af05983627dd94ba217be#diff-333863a57a158a50c87772836c137c99R946):
    * Check whether the sender is allowed to send this
    * Update `Checkable#executions` and broadcast that

**Comment by Al2Klimov** [comment]
@lippserd @N-o-X @htriem I'd very like you **not to merge** any v2.13 stuff before this one.

**Comment by Al2Klimov** [comment]
... or – let's say – before external contributions in general.

**Comment by Al2Klimov** [comment]
@lippserd @N-o-X @htriem @Yonas-net Oh, and please try not to rebase anything unless absolutely neccessary. The one or other commit is already in >1 PR.

@mcodato @bgroeber Don't worry about conflicts with master. I'll handle them. Just keep [this strategy](https://github.com/Icinga/icinga2/pull/8040#issuecomment-637610855) in mind.

**Comment by benjamingroeber** [comment]
> @mcodato @bgroeber Don't worry about conflicts with master. I'll handle them. Just keep [this strategy](https://github.com/Icinga/icinga2/pull/8040#issuecomment-637610855) in mind.

@Al2Klimov Sure, no problem. 
There will at least be 1 other PR regarding the GET endpoint for the results. Maybe we could get going with that one or should we still attend your feedback on https://github.com/Icinga/icinga2/pull/8107 ?

**Comment by Al2Klimov** [comment]
I'd wait for @htriem 's and my feedback.

**Comment by Al2Klimov** [comment]
@mcodato @bgroeber We have the execution process from A to Z, we have the executions attribute for viewing the status via API, we have the cleanup and the docs... I _think_ we have nothing more to implement, do we have?

Anyway the not yet merged PRs need to be reviewed by @htriem before being merged here and this PR needs a full review from @N-o-X before being merged into master. At least the full review won't happen today, now and rather ereyesterday, so if someone already needs this feature, you have to build custom packages by yourself.

**Comment by benjamingroeber** [comment]
Yes, our use case is covered, looking forward to your reviews!

**Comment by Al2Klimov** [comment]
## Test

My last change works:

```
                "executions": {
                    "5ebdd41a-2aec-496e-9407-fa3c173c1947": {
                        "deadline": 1603105281.806762,
                        "end": 1603104981.80678,
                        "exit": 126.0,
                        "output": "Endpoint 's2' doesn't support executing arbitrary commands.",
                        "start": 1603104981.80678
                    },
                    "ecc757b2-4db0-4e58-9c60-3b5bce8387f8": {
                        "deadline": 1603105266.210328,
                        "pending": true
                    }
                },
```

**Comment by Al2Klimov** [comment]
@N-o-X I'd really like this to be reviewed and merged into master (after v2.12.1 stuff of course) before doing anything else with it. Don't worry about the merge conflicts, I'll take care.

File: lib/icinga/apiactions.cpp
Comment (N-o-X) [inline lib/icinga/apiactions.cpp:587]: There should be a newline before each control structure

File: lib/icinga/apiactions.cpp
Comment (N-o-X) [inline lib/icinga/apiactions.cpp:647]: There should be a newline before each control structure

File: lib/icinga/apiactions.cpp
Comment (N-o-X) [inline lib/icinga/apiactions.cpp:652]: There should be a newline before each control structure

File: lib/icinga/apiactions.cpp
Comment (N-o-X) [inline lib/icinga/apiactions.cpp:671]: There should be a newline before each control structure

File: lib/icinga/apiactions.cpp
Comment (N-o-X) [inline lib/icinga/apiactions.cpp:687]: There should be a newline before each control structure

File: lib/icinga/apiactions.cpp
Comment (N-o-X) [inline lib/icinga/apiactions.cpp:698]: There should be a newline before each control structure

File: lib/icinga/apiactions.cpp
Comment (N-o-X) [inline lib/icinga/apiactions.cpp:709]: There should be a newline before each control structure

File: lib/icinga/apiactions.cpp
Comment (N-o-X) [inline lib/icinga/apiactions.cpp:714]: There should be a newline before each control structure

File: lib/icinga/apiactions.cpp
Comment (N-o-X) [inline lib/icinga/apiactions.cpp:726]: There should be a newline before and after each control structure

File: lib/icinga/apiactions.cpp
Comment (N-o-X) [inline lib/icinga/apiactions.cpp:730]: There should be a newline before each control structure

File: lib/icinga/apiactions.cpp
Comment (N-o-X) [inline lib/icinga/apiactions.cpp:742]: There should be a newline before and after each control structure

File: lib/icinga/clusterevents-check.cpp
Comment (N-o-X) [inline lib/icinga/clusterevents-check.cpp:307]: This will just kill Icinga without logging anything. Please at least throw a helpful exception.

**Review by N-o-X — CHANGES_REQUESTED** [review]
Hi guys,

sorry for the long wait. I finally found some time to have a look at your work.

I really like the current implementation of this feature and don't have much to say about the way you implemented it. Although I have requested about 80 changes, most of it is code style and small stuff.

Thanks for the great work and patience!

@Al2Klimov it would be great to have the `executions` API endpoint before merging this. Because without that, this PR doesn't feel complete.

Greetings
Noah

**Comment by Al2Klimov** [comment]
> @Al2Klimov it would be great to have the `executions` API endpoint before merging this. Because without that, this PR doesn't feel complete.

What are you talking about, why that feeling and what would it provide that others don't provide?

File: lib/icinga/clusterevents-check.cpp
Comment (Al2Klimov) [reply lib/icinga/clusterevents-check.cpp:313]: `git diff --ignore-all-space 78a94219aa4470833b89cef784eacc59c869f778..HEAD -- lib/icinga/clusterevents-check.cpp` clearly shows...

```
        } else if (command_type == "event_command") {
+               try {
                        host->ExecuteEventHandler(macros, true);
+               } catch (const std::exception& ex) {
+                       if (params->Contains("source")) {
+                               String output = "Exception occurred while executing event command '" + command + "' for '" +
+                                       host->GetName() + "': " + DiagnosticInformation(ex);
+
+                               double now = Utility::GetTime();
+                               SendEventExecuteCommand(params, ServiceUnknown, output, now, now, listener, origin, sourceEndpoint);
+                       } else {
+                               throw;
+                       }
+               }
```

... that the whole try-catch itself, not just that `throw;` has been added by this PR. So the `throw;` is the old behavior.

I consider this another construction area.

**Comment by N-o-X** [comment]
> > @Al2Klimov it would be great to have the `executions` API endpoint before merging this. Because without that, this PR doesn't feel complete.
> 
> What are you talking about, why that feeling and what would it provide that others don't provide?

Never mind. I think I got confused by those many discussions in all the Issues/PRs. The attribute on checkables should be enough.

**Review by N-o-X — APPROVED** [review]

**Comment by Al2Klimov** [comment]
Finally! 🎉

**Pull request merged**
